parser/auth/sm3.go

```diff
@@ -0,0 +1,384 @@
+		if i%2 == 0 {
+			bufA.Write([]byte(plaintext))
+		} else {
+			bufA.Write(sumB[:])
+		}
+	}
+
+	// 12
+	sumA := SM3(bufA.Bytes())
+	bufA.Reset()
+
+	// 13, 14, 15
+	bufDP := bufA
+	for range []byte(plaintext) {
+		bufDP.Write([]byte(plaintext))
+	}
+	sumDP := SM3(bufDP.Bytes())
+	bufDP.Reset()
+
+	// 16
+	p := make([]byte, 0, sha256.Size)
+	for i = len(plaintext); i > 0; i -= MIXCHARS {
+		if i > MIXCHARS {
+			p = append(p, sumDP[:]...)
+		} else {
+			p = append(p, sumDP[0:i]...)
+		}
+	}
+
+	// 17, 18, 19
+	bufDS := bufA
+	for i = 0; i < 16+int(sumA[0]); i++ {
+		bufDS.Write(salt)
+	}
+	sumDS := SM3(bufDS.Bytes())
+	bufDS.Reset()
+
+	// 20
+	s := make([]byte, 0, 32)
+	for i = len(salt); i > 0; i -= MIXCHARS {
+		if i > MIXCHARS {
+			s = append(s, sumDS[:]...)
+		} else {
+			s = append(s, sumDS[0:i]...)
+		}
+	}
+
+	// 21
+	bufC := bufA
+	var sumC []byte
+	for i = 0; i < iterations; i++ {
+		bufC.Reset()
+		if i&1 != 0 {
+			bufC.Write(p)
+		} else {
+			bufC.Write(sumA[:])
+		}
+		if i%3 != 0 {
+			bufC.Write(s)
+		}
+		if i%7 != 0 {
+			bufC.Write(p)
+		}
+		if i&1 != 0 {
+			bufC.Write(sumA[:])
+		} else {
+			bufC.Write(p)
+		}
+		sumC = SM3(bufC.Bytes())
+		sumA = sumC
+	}
+
+	// 22
+	buf := bytes.NewBuffer(make([]byte, 0, 100))
+	buf.Write([]byte{'$', 'B', '$'})
+	rounds := fmt.Sprintf("%03d", iterations/ITERATION_MULTIPLIER)
+	buf.Write([]byte(rounds))
+	buf.Write([]byte{'$'})
+	buf.Write(salt)
+
+	b64From24bit([]byte{sumC[0], sumC[10], sumC[20]}, 4, buf)
+	b64From24bit([]byte{sumC[21], sumC[1], sumC[11]}, 4, buf)
+	b64From24bit([]byte{sumC[12], sumC[22], sumC[2]}, 4, buf)
+	b64From24bit([]byte{sumC[3], sumC[13], sumC[23]}, 4, buf)
+	b64From24bit([]byte{sumC[24], sumC[4], sumC[14]}, 4, buf)
+	b64From24bit([]byte{sumC[15], sumC[25], sumC[5]}, 4, buf)
+	b64From24bit([]byte{sumC[6], sumC[16], sumC[26]}, 4, buf)
+	b64From24bit([]byte{sumC[27], sumC[7], sumC[17]}, 4, buf)
+	b64From24bit([]byte{sumC[18], sumC[28], sumC[8]}, 4, buf)
+	b64From24bit([]byte{sumC[9], sumC[19], sumC[29]}, 4, buf)
+	b64From24bit([]byte{0, sumC[31], sumC[30]}, 3, buf)
+
+	return buf.String()
+}
+
+// CheckSM3Password checks if a sm3 authentication string matches a password
+func CheckSM3Password(pwhash []byte, password string) (bool, error) {
+	pwhashParts := bytes.Split(pwhash, []byte("$"))
+	if len(pwhashParts) != 4 {
+		return false, errors.New("failed to decode hash parts")
+	}
+
+	hashType := string(pwhashParts[1])
+	if hashType != "B" {
+		return false, errors.New("digest type is incompatible")
+	}
+
+	iterations, err := strconv.Atoi(string(pwhashParts[2]))
+	if err != nil {
+		return false, errors.New("failed to decode iterations")
+	}
+	iterations = iterations * ITERATION_MULTIPLIER
+	salt := pwhashParts[3][:SALT_LENGTH]
+
+	newHash := sm3crypt(password, salt, iterations)
+
+	return bytes.Equal(pwhash, []byte(newHash)), nil
+}
+
+// NewSM3Password creates a new SM3 password hash
+func NewSM3Password(pwd string) string {
```

Comment:
If the only difference between `NewSM3Password` and `NewSha2Password` is the function that it calls, we could combine the code.

```suggestion
func NewHashPassword(pwd string, hashFunction func(string, []bytes) string) string {
```